pymc3_hmm/utils.py

```diff
@@ -172,15 +176,19 @@ def tt_broadcast_arrays(*args):
     """
     p = max(a.ndim for a in args)
 
-    args = [tt.shape_padleft(a, n_ones=p - a.ndim) if a.ndim < p else a for a in args]
+    list_args = [
+        tt.shape_padleft(a, n_ones=p - a.ndim) if a.ndim < p else a for a in args
+    ]
 
```

Comment:
the original error being raised on line 175 was: `Incompatible types in assignment (expression has type "List[Any]", variable has type "Tuple[Any, ...]")`

Initially I tried something like `args: Union[List[Any], Tuple[Any]] = [tt.shape_padleft(...` but that raised an error `Name 'args' already defined on line 164`

Any suggestions on a better approach than this?

Reply:
> Initially I tried something like `args: Union[List[Any], Tuple[Any]] = [tt.shape_padleft(...` but that raised an error `Name 'args' already defined on line 164`

That looks like a potential typo; you might've mean `list_args: Union...`.

Reply:
Also, the typing for `*args` should be something like `*args: TensorVariable`.

Reply:
would TensorType work?

Reply:
I was trying to say that I wasn't able to resolve the mypy error  by typing `args` to a Union, but the mypy error went away when I created the new `list_args` variable;

Reply:
The `Type`s in Aesara/Theano-PyMC are distinct from the objects that represent actual tensors/arrays; those types have the base class `Variable`.  In other words, the objects used to (symbolically) represent actual tensors are `TensorVariable`s (and `TensorConstant`s).  

The [`Type`](https://github.com/pymc-devs/aesara/blob/master/aesara/graph/type.py#L22) objects, like `TensorType`, are actually instantiated and used to pass type information around, as well as perform conversions from/to different `Type`s via `Type.filter`&mdash;for example.  A [`Variable`](https://github.com/pymc-devs/aesara/blob/master/aesara/graph/basic.py#L285) object like `TensorVariable` will carry an instance of a `Type`/`TensorType` in its `Variable.type` attribute, and that's one of the primary connections between those types of objects.

This situation can be a bit confusing, because Aesara is using instances of a `Type` class/type to provide type information, when the actual (i.e. Python) type of an object is usually determined using Python's built-in type system (e.g. `type(var)`).  Simply put, Aesara is implementing its own simple type system.

There is [an open issue](https://github.com/pymc-devs/aesara/issues/134) that discusses this and proposes that we replace these `Type`s with standard Python types.

Reply:
@Veganveins, it looks like your previous comment is missing.  I updated my comment to include it as a quotation.

If you updated or removed the old comment, please refrain from doing so when it substantially changes the comment's content and/or context&mdash;and especially avoid removing such changes from a comment's edit history (unless sensitive information was revealed, or something of similar importance); otherwise, these comment threads become very difficult to follow without referencing the original history (e.g. per email notifications).

Reply:
Is the idea here to do something along the lines of this?

```python
from aesara import TensorVariable
def tt_broadcast_arrays(*args: TensorVariable):
```

I might still be a little confused because it doesn't seem like I can actually import TensorVariable from aesara? Any hints on where TensorVariable comes from and how best to import it to use as a type hint?

Reply:
Yeah, that seems correct, but the import is actually `from aesara.tensor.var import TensorVariable`.